File: android/src/main/java/com/lightningkite/khrysalis/observables/binding/serves.shared.kt
package com.lightningkite.khrysalis.observables.binding

import com.lightningkite.khrysalis.AnyObject
import com.lightningkite.khrysalis.observables.MutableObservableProperty
import com.lightningkite.khrysalis.observables.observable
import com.lightningkite.khrysalis.rx.DisposeCondition
import com.lightningkite.khrysalis.rx.addWeak
import com.lightningkite.khrysalis.rx.until
import io.reactivex.rxkotlin.subscribeBy

@Deprecated("Use something else for binding now")
fun <T> MutableObservableProperty<T>.serves(whilePresent: AnyObject, other: MutableObservableProperty<T>) {

    var suppress = false

    other.observable.addWeak(whilePresent, { ignored, value ->
        if (!suppress) {
            suppress = true
            this.value = value.value
            suppress = false
        }
    })

    this.onChange.addWeak(whilePresent, { ignored, value ->
        if (!suppress) {
            suppress = true
            other.value = value.value
            suppress = false
        }
    })
}

fun <T> MutableObservableProperty<T>.serves(until: DisposeCondition, other: MutableObservableProperty<T>) {

    var suppress = false

    other.observable.subscribeBy { value ->
        if (!suppress) {
            suppress = true
            this.value = value.value
            suppress = false
        }
    }.until(until)

    this.onChange.subscribeBy { value ->
        if (!suppress) {
            suppress = true
            other.value = value.value
            suppress = false
        }
    }.until(until)
}

File: ios/Khrysalis/com/lightningkite/khrysalis/observables/binding/serves.shared.swift
//Package: com.lightningkite.khrysalis.observables.binding
//Converted using Khrysalis2

import Foundation
import Khrysalis
import RxSwift
import RxRelay



extension MutableObservableProperty {
     public func serves(whilePresent: AnyObject, other: MutableObservableProperty<T>) -> Void {
        var suppress = false
        other.observable.addWeak(whilePresent, { (ignored, value) in 
            if !suppress {
                suppress = true
                self.value = value.value
                suppress = false
            }
        })
        self.onChange.addWeak(whilePresent, { (ignored, value) in 
            if !suppress {
                suppress = true
                other.value = value.value
                suppress = false
            }
        })
    }
     public func serves(_ whilePresent: AnyObject, _ other: MutableObservableProperty<T>) -> Void {
        return serves(whilePresent: whilePresent, other: other)
    }
}
 
 

extension MutableObservableProperty {
    public func serves(until: DisposeCondition, other: MutableObservableProperty<T>) -> Void {
        var suppress = false
        other.observable.subscribeBy{ (value) in 
            if !suppress {
                suppress = true
                self.value = value.value
                suppress = false
            }
        }.until(until)
        self.onChange.subscribeBy{ (value) in 
            if !suppress {
                suppress = true
                other.value = value.value
                suppress = false
            }
        }.until(until)
    }
    public func serves(_ until: DisposeCondition, _ other: MutableObservableProperty<T>) -> Void {
        return serves(until: until, other: other)
    }
}
 
 
